examples/frost.c

```diff
@@ -0,0 +1,334 @@
+        }
+        partial_sigs[i] = &signer[signer_id].partial_sig;
+    }
+    /* Communication round 2: A production system would exchange
+     * partial signatures here before moving on. */
+    for (i = 0; i < THRESHOLD; i++) {
+        signer_id = signers[i];
+        /* To check whether signing was successful, it suffices to either verify
+         * the aggregate signature with the aggregate public key using
+         * secp256k1_schnorrsig_verify, or verify all partial signatures of all
+         * signers individually. Verifying the aggregate signature is cheaper but
+         * verifying the individual partial signatures has the advantage that it
+         * can be used to determine which of the partial signatures are invalid
+         * (if any), i.e., which of the partial signatures cause the aggregate
+         * signature to be invalid and thus the protocol run to fail. It's also
+         * fine to first verify the aggregate sig, and only verify the individual
+         * sigs if it does not work.
+         */
+        if (!secp256k1_frost_partial_sig_verify(ctx, &signer[signer_id].partial_sig, &signer[signer_id].pubnonce, &signer[signer_id].pubshare, &signer[signer_id].session, cache)) {
+            return 0;
+        }
+    }
+    return secp256k1_frost_partial_sig_agg(ctx, sig64, &signer[signer_id].session, partial_sigs, THRESHOLD);
+}
+
+int main(void) {
+    secp256k1_context* ctx;
+    int i;
+    struct signer_secrets signer_secrets[N_SIGNERS];
+    struct signer signers[N_SIGNERS];
+    unsigned char sigs[N_SIGNERS][64];
+    secp256k1_xonly_pubkey agg_pk;
+    secp256k1_frost_tweak_cache cache;
+    unsigned char msg[32] = "this_could_be_the_hash_of_a_msg!";
+    unsigned char sig[64];
+
+    /* Create a context for signing and verification */
+    ctx = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
+    printf("Creating key pairs......");
+    for (i = 0; i < N_SIGNERS; i++) {
+        if (!create_keypair(ctx, &signer_secrets[i], &signers[i])) {
+            printf("FAILED\n");
+            return 1;
+        }
+    }
+    printf("ok\n");
+    printf("Creating VSS............");
+    for (i = 0; i < N_SIGNERS; i++) {
+        if (!create_vss(ctx, &signer_secrets[i], &signers[i])) {
+            printf("FAILED\n");
+            return 1;
+        }
+    }
+    printf("ok\n");
+    printf("Creating shares.........");
+    if (!create_shares(ctx, signer_secrets, signers, &agg_pk)) {
+        printf("FAILED\n");
+        return 1;
+    }
+    printf("ok\n");
+    printf("Signing VSS.............");
+    if (!sign_vss(ctx, signer_secrets, signers, sigs)) {
+        printf("FAILED\n");
+        return 1;
+    }
+    printf("ok\n");
+    printf("Verifying VSS...........");
+    for (i = 0; i < N_SIGNERS; i++) {
+        if (!secp256k1_schnorrsig_verify(ctx, sigs[i], signers[0].vss_hash, 32, &signers[i].pubkey)) {
+            printf("FAILED\n");
+            return 1;
+        }
+    }
+    printf("ok\n");
```

Comment:
These signatures, unlike the pok, sign the `vss_hash`, an output of `share_agg`, for the broadcast channel.

Reply:
Is this a *requirement* for the broadcast channel? It seems quite likely to me that the broadcast channel has its own means of ensuring authenticity.

Reply:
Note to self: the `share_agg` doc instructs the user to sign the vss_hash (but doesn't mention secure broadcast).

Reply:
I don't actually think the echo broadcast helps much in that situation because the malicious signer can lie in the echo round just as well to cause abort.

Reply:
> I don't actually think the echo broadcast helps much in that situation because the malicious signer can lie in the echo round just as well to cause abort.

Right... Okay, so I think there are two aspects to this:

First, "abort" is not avoidable. (That is, at least not without much heavier broadcast protocols that then are equivalent to consensus and need honest majority, etc.) An honest signer can get to a state where it doesn't know whether honest signers will succeed or not. You can call this "abort", but it's rather some "indeterminate" output. The good thing about this is: If you get an indeterminate output, then you still know: 
 - If DKG really has not worked, then no other honest signer will get a success output.
 - If DKG has worked, then your secrets are fine, and you just don't know about it because network failures or malicious signers prevented you from learning that it has worked. But if some other signer has used the aggregate key (e.g., has sent money to it), you can in principle participate in a signing session to create signatures (and spend the money). I say "in principle" because you still need to have a FROST session with only honest signers and a working network. So in that case, when you get an indeterminate, you may need to investigate and synchronize out-of-band. 

Now, Protocol 2 in that paper is in that spirit. Either it outputs something (success), or it outputs nothing (indeterminate). But I think your observation is correct. We don't need the full protocol here. In a sense, the second round (the echo round) *alone* is an equality check because its purpose is to agree on what everyone has seen in the first round. So if we just do the echo round, (i.e., everyone sends their VSS hash to everyone, and then everyone checks equality with their local version, and this is what is happening here in the example), this is enough: Assume equality is violated, i.e., two honest signers i and j have different VSS hashes. Then every honest signer will notice and go to "indeterminate" because either the VSS hash received from i or that received from j won't match its local VSS hash.

Reply:
What I was trying to point out is that the second round does not help in establishing agreement ("If an honest party outputs true, then every honest party will output true."), which is stated as a requirement for [SimplPedPop](https://eprint.iacr.org/2023/899.pdf). The echo-broadcast explicitly does not provide agreement (title: "Secure Multi-Party Computation Without Agreement"). Non-agreement becomes a problem when an honest signer thinks the DKG has succeeded and sends money to the address but actually there are honest signers who have aborted due to network conditions or malicious participants.

What I had missed is that due to the nature of the (1-round) echo-broadcast, for any honest signer A who succeeds and honest signer B who aborts, A and B's `vss_hash` is in fact the same. Hence, all honest signers have the same `vss_hash` and the DKG worked in principle. Now the question is if it's possible to repair it and how to do that. A signer that "aborted" should obey to the following rules:

- It should not fail the session or remove the session's secret key material and it should fulfill signing requests (subject to whatever rules had been established).
- Since it's not possible to know whether all participants have shared their keys correctly, the signer that aborted should not send coins to the address derived by the public key.
- If all participants of the DKG signal that they aborted, the signer can forget the session and all its secret key material.

This may be quite annoying in practice, but I don't have a better approach than attempting to resolve this out-of-band if possible.

Reply:
What if, as a final step, the signers run the sign protocol and produce a signature with the key generated by the DKG? A t-of-n signature would demonstrate that t participants can produce a signature, but also an n-of-n signature would demonstrate that all participants have valid DKG outputs.

Reply:
I'm not sure that an n-of-n signature can prove that all participants have valid DKG outputs. It seems possible that both are  true:

- The malicious signer has not shared their key correctly: the vss_hash is different for some honest signers. Therefore, t-of-n are not able to sign (without help from the malicious signer).
- The malicious signer, having drawn the key and sent the incorrect shares, can make n-of-n signing succeed.

Hence, a final (n-of-n) signature would not prove that all participants have valid DKG outputs.

Reply:
Hm, this is how I'd think a malicious signer could make it work in a 2-of-3 setup:

1. Eve generates random key x
2. Eve generates polynomial f for x, sends commitment to f, share and pok to Alice
3. Eve generates polynomial g for x distinct from f, sends commitment to g, share and pok to Bob
4. When creating partial sigs, Alice and Bob use their aggregate shares multiplied with the lagrange coefficient
5. Eve creates its "normal" partial signature, but subtracts Alice and Bobs partial signatures for the garbage shares and adds a partial signature for the key x

This should result in a valid signature

Reply:
But wouldn't this be a signature for key x, not for the output of the DKG, which would be the sum of Alice, Bob, and Eve's keys? If Eve generates random key x, and Alice generates a random key y, and Bob generates a random key z, then how would Eve make the DKG output the key xG for the public key, when it should output (x+y+z)G for the public key? And if the DKG does output (x+y+z)G, then wouldn't Eve need to produce a signature for (x+y+z)G and not simply xG?

Reply:
Sorry, I'm doing a lot of handwaving. It's true that the DKG would have to produce a signature for `x+y+z` and not just `x`. Assume Eve has shares `y'` from Alice and `z'`from Bob. Then the sum of Alice and Bob's partial signature is a signature for `y-y'+z-z'+g` where `g` are the garbage shares sent by Eve. Eve produces a partial sig for `x + y' + z' - g` such that the sum of all partial sigs should be `x+y+z`.

Reply:
`x`, `y`, and `z` are secrets, and `y'` and `z'` are polynomial shares that reconstruct to secrets, and typically shares and secrets are not added to each other. So I'm not following how we can compute a pubkey by adding secrets to shares with `(y-y'+z-z'+g)G` and `(x + y' + z' - g)G`.

Maybe it would help to define a full set of terms:

1. Eve generates random secret `x`, and shares `x_e`, `x_a`, and `x_b`. `x_a` is from polynomial `f` and `x_b` is from polynomial `g`.
2. Alice generates random secret `y`, and shares `y_e`, `y_a`, and `y_b`
3. Bob generates random secret `z`, and shares `z_e`, `z_a`, and `z_b`
4. The DKG pubkey is `P = (x + y + z)G`
5. Eve's aggregate share is `e = x_e + y_e + z_e`
6. Alice's aggregate share is `a = x_a + y_a + z_a`
7. Bob's aggregate share is `b = x_b + y_b + z_b`
8. For the signature, the nonce pubkey is `R = r_e*G + r_a*G + r_b*G`
9. Alice's partial sig is `s_a = r_a + H(P || R || m) * a * l_a`
10. Bob's partial sig is `s_b = r_b + H(P || R || m) * b * l_b`

With this setup, how does Eve produce a valid signature for `(P, R, m)`?